arthur/apis/directory/ldap.py

```diff
@@ -0,0 +1,144 @@
+"""API utilities for fetching data from the LDAP directory."""
+
+from dataclasses import dataclass
+from functools import cache
+
+from bonsai import LDAPClient, LDAPDN, LDAPSearchScope
+
+from arthur.config import CONFIG
+from arthur.constants import LDAP_ROLE_MAPPING
+
+
+@dataclass
+class LDAPUser:
+    """A user in the LDAP directory."""
+
+    uid: str
+    employee_number: str | None = None
+    display_name: str | None = None
+    groups: list[str] | None = None
+
+
+def prepare_dn(dn: str) -> str:
+    """Prepare a DN into a fully qualified DN."""
+    return f"{dn},{CONFIG.ldap_base_dn}"
+
+
+@cache
+def create_client() -> LDAPClient:
+    """Create an LDAP client with the configured settings."""
+    client = LDAPClient(str(CONFIG.ldap_host), tls=True)
+
+    client.set_credentials(
+        "SIMPLE", CONFIG.ldap_bind_user, CONFIG.ldap_bind_password.get_secret_value()
+    )
+    client.set_ca_cert(CONFIG.ldap_certificate_location.as_posix())
+
+    return client
+
+
+async def find_users() -> list[LDAPUser]:
+    """Find all users in the LDAP directory."""
+    client = create_client()
+
+    found_users = []
+
+    async with client.connect(is_async=True) as conn:
+        users = await conn.search(
+            prepare_dn("cn=users,cn=accounts"),
+            LDAPSearchScope.SUBTREE,
+            "(mail=*@pydis.wtf)",
+            ["uid", "employeeNumber", "displayName", "memberOf"],
+        )
+
+        for user in users:
+            groups = user.get("memberOf", None)
+            parsed_groups = []
+
+            if groups:
+                for group in groups:
+                    dn = LDAPDN(group)
+                    g_name = dn.rdns[0][0][1]
+                    if g_name in LDAP_ROLE_MAPPING:
+                        parsed_groups.append(g_name)
+
+            user_data = {
+                "uid": user["uid"][0],
+                "employee_number": user.get("employeeNumber", [None])[0],
+                "display_name": user["displayName"][0],
+                "groups": parsed_groups,
+            }
+
+            found_users.append(LDAPUser(**user_data))
+
+    return found_users
+
+
+async def find_by_discord_id(discord_id: int) -> LDAPUser | None:
+    """Find a user in the LDAP directory by their Discord ID."""
+    client = create_client()
+
+    async with client.connect(is_async=True) as conn:
+        users = await conn.search(
+            prepare_dn("cn=users,cn=accounts"),
+            LDAPSearchScope.SUBTREE,
+            f"(employeeNumber={discord_id})",
+            ["uid", "employeeNumber", "displayName", "memberOf"],
+        )
+
+        user = users[0] if users else None
+
+        if not user:
+            return None
+
+        groups = user.get("memberOf", None)
```

Comment:
If you do `user.get("memberOf", ())`, there's no need for `if groups` :slightly_smiling_face: